File: 백준/Gold/16236. 아기 상어/아기 상어.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, curSize = 2, curX, curY, time = 0, ate = 0;
    static int[][] arr;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, 1, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        arr = new int[n][n];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
                if (arr[i][j] == 9) {
                    curX = i;
                    curY = j;
                    arr[i][j] = 0;
                }
            }
        }

        while (true) {
            int[] fishList = bfs();
            if (fishList == null) break;

            curX = fishList[0];
            curY = fishList[1];
            time += fishList[2];
            arr[curX][curY] = 0;
            ate++;

            if (ate == curSize) {
                curSize++;
                ate = 0;
            }
        }

        System.out.println(time);
    }

    static int[] bfs() {
        ArrayDeque<int[]> q = new ArrayDeque<>();
        q.add(new int[]{curX, curY, 0});

        int minX = -1, minY = -1, minDist = Integer.MAX_VALUE;
        boolean[][] visited = new boolean[n][n];
        visited[curX][curY] = true;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0], y = cur[1], dist = cur[2];

            if (arr[x][y] > 0 && arr[x][y] < curSize) {
                if (dist < minDist || (dist == minDist && (x < minX || (x == minX && y < minY)))) {
                    minX = x;
                    minY = y;
                    minDist = dist;
                }
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dr[i];
                int ny = y + dc[i];

                if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visited[nx][ny] && arr[nx][ny] <= curSize) {
                    visited[nx][ny] = true;
                    q.add(new int[]{nx, ny, dist + 1});
                }
            }
        }

        return minX == -1 ? null : new int[]{minX, minY, minDist};
    }
}

File: 백준/Gold/16236. 아기 상어/아기 상어.py
import heapq
from collections import deque

INF = 1e9
n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]

cur_size = 2
cur_x, cur_y = 0, 0
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

for i in range(n):
    for j in range(n):
        if arr[i][j] == 9:
            cur_x, cur_y = i, j
            arr[cur_x][cur_y] = 0

def bfs():
    q = deque([(cur_x, cur_y, 0)])
    visited = [[False] * n for _ in range(n)]
    visited[cur_x][cur_y] = True
    fish = []

    while q:
        x, y, dist = q.popleft()

        if 1 <= arr[x][y] < cur_size:
            heapq.heappush(fish, (dist, x, y))

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < n and not visited[nx][ny] and arr[nx][ny] <= cur_size:
                visited[nx][ny] = True
                q.append((nx, ny, dist + 1))

    return fish

ans = 0
ate = 0
while True:
    fish_list = bfs()
    if not fish_list:
        print(ans)
        break

    dist, nx, ny = heapq.heappop(fish_list)
    ans += dist
    cur_x, cur_y = nx, ny
    arr[nx][ny] = 0
    ate += 1

    if cur_size == ate:
        cur_size += 1
        ate = 0